### PR TITLE
ANTLR parser

### PR DESCRIPTION
**Pull request by humleflue**

### CODE REVIEW

**Comment by humleflue** [comment]
Jeg har kigget @laurits-mumberg's arbejde igennem, og jeg synes at det ser meget fornuftigt ud :trollface:

File: ANTLR/src/Compiler/SymbolTable/BaseScope.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    protected Map<String, Symbol> symbols = new LinkedHashMap<>();
+
+    //Only use this to create the global scope
+    public BaseScope() { }
+
+    public BaseScope(Scope enclosingScope){
+        setEnclosingScope(enclosingScope);
+    }
+
+    @Override
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/BaseScope.java:18]: We could consider removing @Override tags

File: ANTLR/src/Compiler/SymbolTable/BaseScope.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    public Scope getEnclosingScope() {
+        return enclosingScope;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/BaseScope.java]: Don't we need a this keyword?
```suggestion
        return this.enclosingScope;
```

File: ANTLR/src/Compiler/SymbolTable/BaseScope.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public Symbol getSymbol(String name) {
+        Symbol s = symbols.get(name);
+        if (s != null) {
+            return s;
+        }
+        // if not here, check any enclosing scope
+        if ( enclosingScope != null ) return enclosingScope.getSymbol(name);
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/BaseScope.java]:
```suggestion
        if ( this.enclosingScope != null ) return this.enclosingScope.getSymbol(name);
```

File: ANTLR/src/Compiler/SymbolTable/FunctionType.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    protected List<Type> argumentTypeList;
+
+    public FunctionType(int returnType, List<Type> argumentTypeList){
+        this.returnType = returnType;
+        this.argumentTypeList = argumentTypeList;
+    }
+
+    @Override
+    public int getVal() {
+        return returnType;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/FunctionType.java]:
```suggestion
        return this.returnType;
```

File: ANTLR/src/Compiler/SymbolTable/FunctionType.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    }
+
+    public List<Type> getArgumentTypeList(){
+        return argumentTypeList;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/FunctionType.java]:
```suggestion
        return this.argumentTypeList;
```

File: ANTLR/src/Compiler/SymbolTable/Symbol.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    public String getName(){
+        return this.name;
+    }
+
+    public Type getType() {return this.type; }
+
+    @Override
+    public int hashCode() {
+        return name.hashCode();
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/Symbol.java:20]:
```suggestion
        return this.name.hashCode();
```

File: ANTLR/src/Compiler/SymbolTable/Symbol.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Override
+    public boolean equals(Object obj) {
+        if ( !(obj instanceof Symbol) ) {
+            return false;
+        }
+        if ( obj==this ) {
+            return true;
+        }
+        return name.equals(((Symbol)obj).getName());
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/Symbol.java:31]:
```suggestion
        return this.name.equals(((Symbol)obj).getName());
```

File: ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.tree.ParseTreeProperty;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class SymbolDefListener extends LangBaseListener{
+    public ParseTreeProperty<Scope> scopes = new ParseTreeProperty<Scope>();
+    public BaseScope globalScope = new BaseScope();
+    Scope currentScope;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java:14]: Why is there no access modifier here?

File: ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public void enterProg(ProgContext ctx) {
+        currentScope = globalScope;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java:18]:
```suggestion
        this.currentScope = globalScope;
```

File: ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void enterFuncdef(FuncdefContext ctx) {
+        //Gets lists of parameters as ParseRuleContexts
+        List<FuncdefparamContext> params =  ctx.getRuleContext(FuncdefparamsContext.class, 0)
+                .getRuleContexts(FuncdefparamContext.class);
+
+        //Might be usefull for type-checking. Delete otherwise
+        List<Type> argumentList = new ArrayList<Type>();
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java:28]: It is a bit confusing that our Type is called Type. As one might think that Type is a generic keyword. But it is actually a class in our implementation.

File: ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        currentScope.defineSymbol(new Symbol(ctx.ID().getText(), type));
+
+        Scope newScope = new BaseScope(currentScope);
+        attachScope(ctx, newScope);
+        currentScope = newScope;
+    }
+
+    @Override
+    public void exitFuncdef(FuncdefContext ctx) {
+        currentScope = currentScope.getEnclosingScope();
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java:43]:
```suggestion
        this.currentScope = currentScope.getEnclosingScope();
```

File: ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public void exitFuncdefparam(FuncdefparamContext ctx) {
+        Type paramType = new BaseType(ctx.start.getType());
+        Symbol paramSymbol = new Symbol(ctx.ID().getText(), paramType);
+        currentScope.defineSymbol(paramSymbol);
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolDefListener.java:50]:
```suggestion
       this.currentScope.defineSymbol(paramSymbol);
```

File: ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
+package Compiler.SymbolTable;
+
+import Compiler.Lang.LangBaseListener;
+import Compiler.Lang.LangParser.*;
+import org.antlr.v4.runtime.tree.ParseTreeProperty;
+
+public class SymbolRefListener extends LangBaseListener{
+    ParseTreeProperty<Scope> scopes = new ParseTreeProperty<Scope>();
+    Scope globalScope;
+    Scope currentScope;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java:10]: Access modifiers?

File: ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    public SymbolRefListener(BaseScope globalScope, ParseTreeProperty<Scope> scopes) {
+        this.scopes = scopes;
+        this.globalScope = globalScope;
+    }
+
+    @Override
+    public void enterProg(ProgContext ctx) {
+        currentScope = globalScope;
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java:19]:
```suggestion
        this.currentScope = globalScope;
```

File: ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public void enterFuncdef(FuncdefContext ctx) {
+        currentScope = scopes.get(ctx);
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java:24]:
```suggestion
        this.currentScope = scopes.get(ctx);
```

File: ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    public void exitFuncdef(FuncdefContext ctx) {
+        currentScope = currentScope.getEnclosingScope();
```
Comment (JacobToftgaardRasmussen) [inline ANTLR/src/Compiler/SymbolTable/SymbolRefListener.java:29]:
```suggestion
        this.currentScope = currentScope.getEnclosingScope();
```

File: CUP-parser/AST/Abstract/BinaryOp.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package AST.Abstract;
+
+import AST.Master.Node;
+
+public abstract class BinaryOp implements Node{
+    public BinaryOp(Node left, Node right, String operator) {
+        Left = left;
+        Right = right;
+        Operator = operator;
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Abstract/BinaryOp.java]:
```suggestion
        this.Left = left;
        this.Right = right;
        this.Operator = operator;
```

File: CUP-parser/AST/Abstract/BinaryOp.java [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+    }
+
+    public String NodeType;
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Abstract/BinaryOp.java]: How come NodeType is not set in the constructor?

File: CUP-parser/AST/Nodes/Followterm.java [hunk 19]
```diff
@@ -0,0 +1,14 @@
+package AST.Nodes;
+
+import AST.Master.Node;
+
+public class Followterm implements Node {
+    public Followterm(Simpleops simpleops, Math math) {
+        Simpleops = simpleops;
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Nodes/Followterm.java]:
```suggestion
        this.Simpleops = simpleops;
```

File: CUP-parser/AST/Nodes/Math.java [hunk 20]
```diff
@@ -0,0 +1,14 @@
+package AST.Nodes;
+
+import AST.Abstract.*;
+
+public class Math extends Stmt {
+    public Math(AST.Abstract.Term term, AST.Nodes.Followterm followterm) {
+        Term = term;
+        Followterm = followterm;
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Nodes/Math.java]:
```suggestion
        this.Term = term;
        this.Followterm = followterm;
```

File: CUP-parser/AST/Nodes/Simpleops.java [hunk 21]
```diff
@@ -0,0 +1,8 @@
+package AST.Nodes;
+
+import AST.Abstract.BinaryOp;
+
+public class Simpleops extends BinaryOp {
+    public Simpleops() {
+    }
+}
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Nodes/Simpleops.java]:
```suggestion
    }
}

```

Comment (JacobToftgaardRasmussen) [reply CUP-parser/AST/Nodes/Simpleops.java]: Missing newline at end of file, which git likes.

File: CUP-parser/AST/Nodes/TextVal.java [hunk 22]
```diff
@@ -0,0 +1,12 @@
+package AST.Nodes;
+
+import AST.Abstract.*;
+
+public class TextVal extends Terminal {
+    public TextVal(String spelling) {
+        Spelling = spelling;
```
Comment (JacobToftgaardRasmussen) [inline CUP-parser/AST/Nodes/TextVal.java]:
```suggestion
        this.Spelling = spelling;
```

File: CUP-parser/AST/Abstract/BinaryOp.java (hunk 18 above)
Comment (laurits-mumberg) [reply CUP-parser/AST/Abstract/BinaryOp.java]: @christopher-jensen

File: ANTLR/src/Compiler/SymbolTable/BaseScope.java (hunk 1 above)
Comment (laurits-mumberg) [reply ANTLR/src/Compiler/SymbolTable/BaseScope.java:18]: tester jeg lige

**Review by JacobToftgaardRasmussen — COMMENTED** [review]
Looks good

**Review by JacobToftgaardRasmussen — APPROVED** [review]
Now I got it right

**Comment by laurits-mumberg** [comment]
:trollface:

**Comment by humleflue** [comment]
:trollface:

**Pull request merged**